File: ulyp-ui/src/main/java/com/ulyp/ui/renderers/RenderedCollection.java
package com.ulyp.ui.renderers;

import com.ulyp.core.printers.CollectionRepresentation;
import com.ulyp.ui.RenderSettings;
import com.ulyp.ui.util.StyledText;
import javafx.scene.Node;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static com.ulyp.ui.util.CssClass.*;

public class RenderedCollection extends RenderedObject {

    protected RenderedCollection(CollectionRepresentation representation, RenderSettings renderSettings) {
        super(representation.getType());

        List<RenderedObject> renderedObjects = representation.getRecordedItems()
                .stream()
                .map(repr -> RenderedObject.of(repr, renderSettings))
                .collect(Collectors.toList());

        List<Node> texts = new ArrayList<>();

        if (renderSettings.showTypes()) {
            texts.add(StyledText.of(representation.getType().getName(), CALL_TREE_TYPE_NAME));
            texts.add(StyledText.of(": ", CALL_TREE_PLAIN_TEXT));
        }

        texts.add(StyledText.of("{", CALL_TREE_COLLECTION_BRACE));

        for (int i = 0; i < renderedObjects.size(); i++) {
            texts.add(renderedObjects.get(i));
            if (i != renderedObjects.size() - 1 || renderedObjects.size() < representation.getLength()) {
                texts.add(StyledText.of(", ", CALL_TREE_PLAIN_TEXT));
            }
        }

        if (renderedObjects.size() < representation.getLength()) {
            texts.add(StyledText.of((representation.getLength() - renderedObjects.size()) + " more...", CALL_TREE_PLAIN_TEXT));
        }

        texts.add(StyledText.of("}", CALL_TREE_COLLECTION_BRACE));

        super.getChildren().addAll(texts);
    }
}

File: ulyp-ui/src/main/kotlin/com/ulyp/ui/renderers/RenderedCollection.kt
package com.ulyp.ui.renderers

import com.ulyp.core.recorders.CollectionRecord
import com.ulyp.core.recorders.ObjectRecord
import com.ulyp.ui.RenderSettings
import com.ulyp.ui.util.CssClass
import com.ulyp.ui.util.StyledText.of
import javafx.scene.Node
import java.util.stream.Collectors

class RenderedCollection(record: CollectionRecord, renderSettings: RenderSettings) : RenderedObject(record.type) {

    init {
        val renderedObjects = record.recordedItems
            .stream()
            .map { record: ObjectRecord -> of(record, renderSettings) }
            .collect(Collectors.toList())
        val texts: MutableList<Node> = ArrayList()
        if (renderSettings.showTypes()) {
            texts.add(of(record.type.name, CssClass.CALL_TREE_TYPE_NAME))
            texts.add(of(": ", CssClass.CALL_TREE_NODE_SEPARATOR))
        }
        texts.add(of("{", CssClass.CALL_TREE_COLLECTION_BRACKET))
        for (i in renderedObjects.indices) {
            texts.add(renderedObjects[i])
            if (i != renderedObjects.size - 1 || renderedObjects.size < record.length) {
                texts.add(of(", ", CssClass.CALL_TREE_NODE_SEPARATOR))
            }
        }
        if (renderedObjects.size < record.length) {
            texts.add(
                of(
                    (record.length - renderedObjects.size).toString() + " more...",
                    CssClass.CALL_TREE_NODE_SEPARATOR
                )
            )
        }
        texts.add(of("}", CssClass.CALL_TREE_COLLECTION_BRACKET))
        super.getChildren().addAll(texts)
    }
}